presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -79,31 +91,163 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
+            }
+
+            // Rewrite to aggregation on symbols which we know are constants and are also part of output
+            AggregationNode aggregationNode = new AggregationNode(
+                    idAllocator.getNextId(),
+                    node.getSource(),
+                    ImmutableMap.of(),
+                    ImmutableList.of(ImmutableList.<Symbol>builder()
+                            .addAll(constantSymbols)
+                            .addAll(node.getOutputSymbols())
+                            .build()),
+                    AggregationNode.Step.SINGLE,
+                    Optional.empty(),
+                    Optional.empty());
+
+            childDecorrelationResultOptional = aggregationNode.accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent() || !childDecorrelationResultOptional.get().atMostSingleRow) {
+                return Optional.empty();
+            }
+
+            return childDecorrelationResultOptional;
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitEnforceSingleRow(EnforceSingleRowNode node, Void context)
+        {
+            Optional<DecorrelationResult> childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent() || !childDecorrelationResultOptional.get().atMostSingleRow) {
+                return Optional.empty();
+            }
+
+            return childDecorrelationResultOptional;
+        }
+
+        @Override
+        public Optional<DecorrelationResult> visitAggregation(AggregationNode node, Void context)
+        {
+            if (node.hasEmptyGroupingSet()) {
+                return Optional.empty();
+            }
+
+            Optional<DecorrelationResult> childDecorrelationResultOptional = lookup.resolve(node.getSource()).accept(this, null);
+            if (!childDecorrelationResultOptional.isPresent()) {
+                return Optional.empty();
+            }
+
+            DecorrelationResult childDecorrelationResult = childDecorrelationResultOptional.get();
+            Set<Symbol> constantSymbols = childDecorrelationResult.getConstantSymbols();
+
+            AggregationNode decorrelatedAggregation = childDecorrelationResult.getCorrelatedSymbolMapper()
+                    .map(node, childDecorrelationResult.node);
+
+            ImmutableList.Builder<List<Symbol>> newGroupingSets = ImmutableList.builder();
+            for (List<Symbol> groupingSet : decorrelatedAggregation.getGroupingSets()) {
+                Set<Symbol> set = ImmutableSet.copyOf(groupingSet);
+                List<Symbol> symbolsToAdd = childDecorrelationResult.symbolsToPropagate.stream()
+                        .filter(symbol -> !set.contains(symbol))
+                        .collect(toImmutableList());
+
+                if (!constantSymbols.containsAll(symbolsToAdd)) {
+                    return Optional.empty();
```

Comment:
how is that possible? `symbolsToAdd` are derived from correlation, and `constantSymbols` are also derived from correlation. Why they could differ? Maybe there is a bug somewhere.

Reply:
If you have predicate, `corr_sym > sym_b`.
Then `symbolsToAdd=[sym_b]`, but `constantSymbols:=[]`.

For predicate, `corr_sym = sym_b`.
Then `symbolsToAdd=[sym_b]`, but `constantSymbols:=[sym_b]`.